// better-call/dist/endpoint.mjs
import { APIError, BetterCallError, ValidationError } from "./error.mjs";
import { isAPIError, tryCatch } from "./utils.mjs";
import { createInternalContext } from "./context.mjs";
import { toResponse } from "./to-response.mjs";

//#region src/endpoint.ts
function createEndpoint(pathOrOptions, handlerOrOptions, handlerOrNever) {
	const path = typeof pathOrOptions === "string" ? pathOrOptions : void 0;
	const options = typeof handlerOrOptions === "object" ? handlerOrOptions : pathOrOptions;
	const handler = typeof handlerOrOptions === "function" ? handlerOrOptions : handlerOrNever;
	if ((options.method === "GET" || options.method === "HEAD") && options.body) throw new BetterCallError("Body is not allowed with GET or HEAD methods");
	if (path && /\/{2,}/.test(path)) throw new BetterCallError("Path cannot contain consecutive slashes");
	const runtimeOptions = {
		method: options.method,
		body: options.body,
		query: options.query,
		error: options.error,
		requireHeaders: options.requireHeaders,
		requireRequest: options.requireRequest,
		cloneRequest: options.cloneRequest,
		disableBody: options.disableBody,
		metadata: options.metadata,
		use: options.use,
		onAPIError: options.onAPIError,
		onValidationError: options.onValidationError
	};
	const internalHandler = async (...inputCtx) => {
		const context = inputCtx[0] || {};
		const { data: internalContext, error: validationError } = await tryCatch(createInternalContext(context, {
			options: runtimeOptions,
			path
		}));
		if (validationError) {
			if (!(validationError instanceof ValidationError)) throw validationError;
			if (options.onValidationError) await options.onValidationError({
				message: validationError.message,
				issues: validationError.issues
			});
			throw new APIError(400, {
				message: validationError.message,
				code: "VALIDATION_ERROR"
			});
		}
		const response = await handler(internalContext).catch(async (e) => {
			if (isAPIError(e)) {
				const onAPIError = options.onAPIError;
				if (onAPIError) await onAPIError(e);
				if (context.asResponse) return e;
			}
			throw e;
		});
		const responseHeaders = internalContext.responseHeaders;
		const status = internalContext.responseStatus;
		return context.asResponse ? toResponse(response, {
			headers: responseHeaders,
			status
		}) : context.returnHeaders ? context.returnStatus ? {
			headers: responseHeaders,
			response,
			status
		} : {
			headers: responseHeaders,
			response
		} : context.returnStatus ? {
			response,
			status
		} : response;
	};
	internalHandler.options = runtimeOptions;
	internalHandler.path = path;
	return internalHandler;
}
createEndpoint.create = (opts) => {
	return (path, options, handler) => {
		return createEndpoint(path, {
			...options,
			use: [...options?.use || [], ...opts?.use || []]
		}, handler);
	};
};

//#endregion
export { createEndpoint };
