// better-call/dist/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
const require_error = require('./error.cjs');
const require_cookies = require('./cookies.cjs');
const require_to_response = require('./to-response.cjs');
const require_endpoint = require('./endpoint.cjs');
const require_middleware = require('./middleware.cjs');
const require_openapi = require('./openapi.cjs');
const require_router = require('./router.cjs');

exports.APIError = require_error.APIError;
exports.BetterCallError = require_error.BetterCallError;
exports.ValidationError = require_error.ValidationError;
exports.createEndpoint = require_endpoint.createEndpoint;
exports.createMiddleware = require_middleware.createMiddleware;
exports.createRouter = require_router.createRouter;
exports.generateOpenAPI = require_openapi.generator;
exports.getOpenAPIHTML = require_openapi.getHTML;
exports.kAPIErrorHeaderSymbol = require_error.kAPIErrorHeaderSymbol;
exports.serializeSignedCookie = require_cookies.serializeSignedCookie;
exports.statusCodes = require_error.statusCodes;
exports.toResponse = require_to_response.toResponse;
